test/test_torch.py

```diff
@@ -17838,407 +17839,435 @@ def inner(self, device, dtype):
-                           float_precision,
-                           dtype_list,
-                           dtype_cpu_list,
-                           decorators):
+def generate_test_function(cls, testMeta):
+
     def fn(self, device, dtype):
         # Generates the CPU inputs
         # Note: CPU tensors are never torch.half
-        cpu_tensor = tensor_ctor(dtype, 'cpu')
-        cpu_args = arg_ctor(dtype, 'cpu')
+        cpu_tensor = testMeta.ctor(dtype, 'cpu')
+        cpu_args = testMeta.arg_ctor(dtype, 'cpu')
 
         # Converts CPU tensors to device tensors
         device_tensor = cpu_tensor.to(dtype=dtype, device=device)
         device_args = [arg.to(device=device) if isinstance(arg, torch.Tensor) else arg for arg in cpu_args]
 
+        rtol = None
+        atol = None
         # Converts float device tensors to half/bfloat16 when the dtype is half/bfloat16
         # Note: CPU half tensors don't support many operations.
         if dtype in {torch.half, torch.bfloat16}:
             device_args = [arg.to(dtype=dtype) if
                            (isinstance(arg, torch.Tensor) and arg.dtype == torch.float) else arg
                            for arg in device_args]
+            (rtol, atol) = testMeta.half_tol
+        elif dtype == torch.float32 and os.name == 'nt':
+            # TestTensorDeviceOpsCUDA.test_qr_big_cuda_float32 has lower tolerance on windows.
+            # todo: investigate platform difference
+            (rtol, atol) = testMeta.win_float_tol
 
         # Runs the tensor op on CPU and device
-        cpu_result = getattr(cpu_tensor, op_str)(*cpu_args)
-        device_result = getattr(device_tensor, op_str)(*device_args)
-
-        dtype2precision = {torch.half : half_precision,
-                           torch.bfloat16 : bfloat16_precision}
-
-        # Compares CPU and device inputs and outputs
-        precision = dtype2precision.get(dtype, float_precision)
+        cpu_result = getattr(cpu_tensor, testMeta.op_name)(*cpu_args)
+        device_result = getattr(device_tensor, testMeta.op_name)(*device_args)
 
-        self.assertEqual(cpu_tensor, device_tensor, atol=precision, exact_dtype=False)
-        self.assertEqual(cpu_args, device_args, atol=precision, exact_dtype=False)
-        self.assertEqual(cpu_result, device_result, atol=precision, exact_dtype=False)
+        self.assertEqual(device_tensor, cpu_tensor, rtol=rtol, atol=atol, exact_dtype=False)
+        self.assertEqual(device_args, cpu_args, rtol=rtol, atol=atol, exact_dtype=False)
+        self.assertEqual(device_result, cpu_result, rtol=rtol, atol=atol, exact_dtype=False)
 
-    test_name = "test_" + op_str + subtest_str
+    test_name = testMeta.test_name()
     assert not hasattr(cls, test_name), "{0} already in TestDevicePrecision".format(test_name)
 
     # Constructs decorator list and applies decorators
-    if decorators is None:
-        decorators = [dtypes(*dtype_list)]
+    if testMeta.decorators is None:
```

Comment:
afaict, `testMeta.decorators` is never None. It's always a list (and can be empty)